### PR TITLE
feat(jellyfish-api-core): implement refundHtlcAll

### PR DESCRIPTION
**Pull request by cwkang1998**
#### What this PR does / why we need it:
Implements spv `spv_refundhtlcall` rpc call.

#### Which issue(s) does this PR fixes?:

Fixes part of #202 

#### Additional comments?:
Need some confirmation with how the `ain` implementation of this rpc is. Right now `ain`'s implementation returns an array of strings, containing only one string, shown [here](https://github.com/DeFiCh/ain/blob/86f554f2454edc8c16684cac6f7332b437a991e0/src/spv/spv_wrapper.cpp#L1354-L1379).

The [rpc documentation from `ain`](https://github.com/DeFiCh/ain/blob/86f554f2454edc8c16684cac6f7332b437a991e0/src/spv/spv_rpc.cpp#L1126) however shows that it returns a string txid, which is not the case.

This current implementation for jellyfish deals with this by simply taking the first element of the returned array.

Edit: This rpc now returns a string array, faithful to its `ain` implementation.

### CODE REVIEW

**Review by canonbrother — COMMENTED** [review]
LGTM!!
(Regarding the **return arr** from ain.. if that does really matter.. we'd open issue on ain)

**Comment by ivan-zynesis** [comment]
> LGTM!! (Regarding the **return arr** from ain.. if that does really matter.. we'd open issue on ain)

Hmm prefer to return array instead for consistencies. I am thinking if ain side is a bug, what if the wallet has multiple htlc, then it should return txid of each refunded.

**Review by canonbrother — DISMISSED** [review]

**Comment by cwkang1998** [comment]
> > LGTM!! (Regarding the **return arr** from ain.. if that does really matter.. we'd open issue on ain)
> 
> Hmm prefer to return array instead for consistencies. I am thinking if ain side is a bug, what if the wallet has multiple htlc, then it should return txid of each refunded.

If so then I think I'll make it return a string array, but I'll just note it somewhere that there is this inconsistency then.

**Review by ivan-zynesis — APPROVED** [review]

**Review by canonbrother — APPROVED** [review]

File: packages/jellyfish-api-core/__tests__/category/spv/refundHtlcAll.test.ts [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    const htlc = await container.call('spv_createhtlc', [pubKeyA, pubKeyB, '10'])
+
+    await testing.container.call('spv_sendtoaddress', [htlc.address, 0.00000546]) // Funds HTLC address with dust
+
+    const promise = testing.rpc.spv.refundHtlcAll(await container.call('spv_getnewaddress'))
+    await expect(promise).rejects.toThrow(RpcApiError)
+    await expect(promise).rejects.toThrow('No unspent HTLC outputs found')
+  })
+
+  it('should not refundHtlcAll when not htlc is created for wallet', async () => {
```
Comment (surangap) [inline packages/jellyfish-api-core/__tests__/category/spv/refundHtlcAll.test.ts:214]:
```suggestion
  it('should not refundHtlcAll when no htlc is created for wallet', async () => {
```

File: packages/jellyfish-api-core/__tests__/category/spv/refundHtlcAll.test.ts [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    /**
+     * Assert that the destination address received the refund
+     */
+    const listReceivingAddresses: any[] = await container.call('spv_listreceivedbyaddress')
+    const receivingAddress = listReceivingAddresses.find(({ address }: { address: string }) => address === destinationAddress)
+    expect(receivingAddress.address).toStrictEqual(destinationAddress)
+    expect(receivingAddress.txids.some((txid: string) => txid === results[0])).toStrictEqual(true)
+  })
+
+  it('should refundHtlcAll for expired only', async () => {
```
Comment (surangap) [inline packages/jellyfish-api-core/__tests__/category/spv/refundHtlcAll.test.ts:79]: do you expect `htlc2` to be no refunded in this test? wondering where is the `for expired only`
Refer https://github.com/DeFiCh/jellyfish/pull/529#discussion_r688293825

Comment (cwkang1998) [reply packages/jellyfish-api-core/__tests__/category/spv/refundHtlcAll.test.ts:79]: Yes I do. In real life scenario, `refundHtlc` would probably not be called before timeout. However in this case, `refundHtlcAll` would be called in a real life scenario, as it initiate refunds only for those that already timeout, and not htlc that has not timed out yet.

Comment (cwkang1998) [reply packages/jellyfish-api-core/__tests__/category/spv/refundHtlcAll.test.ts:79]: Should I change the title to indicate for timeout?

**Review by ivan-zynesis — APPROVED** [review]

**Review by surangap — APPROVED** [review]

**Review by canonbrother — APPROVED** [review]

**Pull request merged**